### PR TITLE
New feature : Library

### LINKED ISSUE #64: New Feature: Library
**Is your feature request related to a problem? Please describe.**
A new page having a list of already read books by the user.

**Describe the solution you'd like**
I would like to create a new page which has a list of all the already rated books by the user which can be accessed through the navbar, the theme of the page would be consistent with the website.

**Are you willing to contribute to this issue?** Yes, I am willing to contribute to this issue as a participant in Gssoc'21.

**Describe alternatives you've considered**
None as of yet.

**Additional context**
As an avid reader myself I feel that having a page with books that have been already read by the user is an important feature.

### PR DESCRIPTION
**Pull request by Fairspy**
## Description
- Added the library page showing the already rated books of the user
- Fixes #64 

## Affected Dependencies
No dependencies were affected

## How has this been tested?
All the previously, as well as newly defined tests, were passed 

## GIF Before
- 

Before

## GIF After

After(3)

After(4)

(I couldn't add the whole after GIF together as the file size was too big)
## Checklist
- [X] I have followed the [Contribution Guidelines](https://github.com/Praful932/Kitabe/blob/master/CONTRIBUTING.md).
- [X] My changes do not edit/add any unrequired files.
- [X] My changes are covered by tests.

### CODE REVIEW

File: BookRecSystem/settings.py [hunk 1]
```diff
@@ -24,6 +24,7 @@
 # SECURITY WARNING: keep the secret key used in production secret!
 SECRET_KEY = os.environ.get('KITABE_SECRET_KEY')
```
Comment (Praful932) [inline BookRecSystem/settings.py:25]: Unstage this file..

File: mainapp/helpers.py [hunk 2]
```diff
@@ -74,7 +74,7 @@ def get_rated_bookids(user_ratings):
     '''
     already_rated = []
     for rating in user_ratings:
-        book_id = rating.bookrating
+        book_id = rating.bookid
```
Comment (Praful932) [inline mainapp/helpers.py:77]: Nice catch, so then recommendation was a bit bugged earlier xD
@maneprajakta See this [Line](https://github.com/Praful932/Kitabe/blob/master/mainapp/views.py#L67)

File: mainapp/tests.py [hunk 3]
```diff
@@ -209,3 +209,36 @@ def test_user_rated_book_valid(self):
         self.assertEquals(rating.bookrating, valid_bookrating)
         self.assertEquals(rating.user, self.user)
         self.client.logout()
+
+
+class RatedBooksTestCase(TestCase):
+    '''
+    Read books view test case
+    '''
```
Comment (Praful932) [inline mainapp/tests.py:217]:
```suggestion
    """Already Read Books View Test Case"""
```

Comment (Praful932) [reply mainapp/tests.py:217]: Would suggest to follow this for every function, we are currently updating our docs #36

File: mainapp/views.py [hunk 4]
```diff
@@ -79,3 +80,26 @@ def book_recommendations(request):
     else:
         return redirect('index')
     return render(request, 'mainapp/recommendation.html', {'books': all_books_dict})
+
+
+@login_required
+@ensure_csrf_cookie
+def read_books(request):
```
Comment (Praful932) [inline mainapp/views.py:87]: Same here, follow numpy doc style for writing docstrings

**Review by Praful932 — CHANGES_REQUESTED** [review]
I have added some comments, please resolve them

File: mainapp/helpers.py (hunk 2 above)
Comment (maneprajakta) [reply mainapp/helpers.py:77]: Yep seems now better recommendation we will get :)

File: static/mainapp/css/read.css [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+  .card-title {
+    font-family: "Sora", sans-serif;
+    font-size: 2rem;
+    color: #d7722c;
+    padding-bottom: 20px;
+  }
+
+  .card-text {
+    font-family: "Pacifico", cursive;
```
Comment (maneprajakta) [inline static/mainapp/css/read.css:39]: font-family for card-text : "Courgette", cursive;
We are following this.

**Review by maneprajakta — CHANGES_REQUESTED** [review]
Requested Changes.

**Comment by Fairspy** [comment]
@Praful932 @maneprajakta I've added the requested changes please review them.
Also, I tried to unstage settings.py but I'm not sure if it worked or not.

**Review by Praful932 — CHANGES_REQUESTED** [review]
The file is still staged :)
Follow [this](https://stackoverflow.com/questions/38743912/remove-a-file-from-a-git-pull-request)

**Comment by Fairspy** [comment]
@Praful932 I've tried to unstage settings.py let me know if it worked or not.

**Comment by Praful932** [comment]
@Fairspy Yep it's unstaged, There are some changes requested in the conversation, We haved added our reviewes, please resolve them

**Comment by Fairspy** [comment]
@Praful932 I've added the requested changes please review them.
Let me know if any further changes are required.

**Comment by Praful932** [comment]
@Fairspy Please check the Files changed to resolve issues, there are still [some](https://github.com/Praful932/Kitabe/pull/100/files#r594573902) pending

**Comment by Fairspy** [comment]
@Praful932 Is it okay?
I went through the numpy style guide but I am not completely sure about how to write docstrings for views if you need me to add something else please let me know.

**Comment by Praful932** [comment]
@Fairspy for now we have decided to keep docs on low prio, we'll make a new issue for that specifically
Please check [this](https://github.com/Praful932/Kitabe/pull/100#discussion_r594830970)

File: mainapp/urls.py [hunk 6]
```diff
@@ -5,6 +5,7 @@
     path('genre_books/<genre>', views.genre_books, name='genre_books'),
     path('explore_books/', views.explore_books, name='explore_books'),
     path('book_recommendations/', views.book_recommendations, name='book_recommendations'),
+    path('read_books/', views.read_books, name='read_books')
```
Comment (Praful932) [inline mainapp/urls.py:8]: let's make this `library`

**Review by Praful932 — CHANGES_REQUESTED** [review]

**Comment by Fairspy** [comment]
@Praful932 I've changed the path to `library` and I had changed the font of `card-text` to `courgette` earlier.
Let me know if any further changes are required.

**Review by maneprajakta — CHANGES_REQUESTED** [review]
- Fix Rate this button as it's on other pages.
Screenshot from 2021-03-19 22-04-54
- Increase font size of book author

**Comment by Praful932** [comment]
@maneprajakta I'm thinking on the lines of how we should structure the library
- Library -> read_books
- Library -> to_read
then their path could be like this `/library/read_books` (This one)
& `/library/to_read`
Wdyt

**Comment by maneprajakta** [comment]
> @maneprajakta I'm thinking on the lines of how we should structure the library
> 
> * Library -> read_books
> * Library -> to_read
>   then their path could be like this `/library/read_books` (This one)
>   & `/library/to_read`
>   Wdyt

Seems great to me :+1: 
In library page itself a left corner we provide both option and by default it would be read_books

**Review by Praful932 — CHANGES_REQUESTED** [review]
image
Button width needs to be fixed

**Comment by Fairspy** [comment]
@Praful932 I've fixed the button width you can check it out.
@maneprajakta I've increased the font-size of book author let me know if you want it changed further.
As for the behaviour of the rate book button I can't really see the same behaviour on my end so I don't really know how to go about fixing the issue are you seeing it for all the books or is it specifically for that one?

**Comment by Fairspy** [comment]
I'm really sorry I force pushed to the branch and it deleted your previous commit I've committed the changes to urls.py again I'm really sorry.

**Review by Praful932 — APPROVED** [review]
Seems good to me now

**Comment by Praful932** [comment]
> I'm really sorry I force pushed to the branch and it deleted your previous commit I've committed the changes to urls.py again I'm really sorry.

Np, the rate button seems fixed now

**Review by maneprajakta — APPROVED** [review]

**Comment by Praful932** [comment]
After revisting this @maneprajakta  Since this section is of already rated books, so instead of displaying `Rate this book` on every book, why don't we display the user's rating directly, so user does not confused

**Comment by maneprajakta** [comment]
@Praful932 
yep a valid point.
I think we can create a new issue for this one.

**Review by Praful932 — APPROVED** [review]
Splendid Work @Fairspy This marks our 100th PR! 🚀

File: static/mainapp/css/read.css (hunk 5 above)
Comment (Praful932) [reply static/mainapp/css/read.css]: @maneprajakta Is this pending?

Comment (maneprajakta) [reply static/mainapp/css/read.css]: @Praful932  Nope

**Pull request merged**